File: js/HUD.js
/**
 * HUD.js — In-game overlay: damage %, stocks, ultimate meter,
 *           shield HP, character name, match timer.
 */
(function () {
const S = SMASH.Settings;

class HUD {
    constructor() {
        this._frame = 0;
    }

    /**
     * @param {CanvasRenderingContext2D} ctx
     * @param {Array} players — [{port, fighter, controller, isAI}]
     * @param {number} [matchTime] — elapsed seconds
     * @param {object} [modeInfo] — { gameMode, waveNumber, waveEnemiesLeft }
     */
    render(ctx, players, matchTime, modeInfo) {
        this._frame++;
        const n = players.length;
        if (n === 0) return;
        const mode = (modeInfo && modeInfo.gameMode) || 'stock';

        // Match timer (top-center)
        if (matchTime !== undefined) {
            this._renderTimer(ctx, matchTime);
        }

        // Wave defense header
        if (mode === 'wave' && modeInfo) {
            this._renderWaveHeader(ctx, modeInfo.waveNumber, modeInfo.waveEnemiesLeft);
        }

        // Draft mode header
        if (mode === 'draft' && modeInfo && modeInfo.draftRemaining) {
            this._renderDraftHeader(ctx, modeInfo.draftRemaining);
        }

        // Filter: in wave mode, only show human player panels
        const displayPlayers = mode === 'wave'
            ? players.filter(p => !p.isWaveEnemy)
            : players;

        // Player panels (bottom)
        const dn = displayPlayers.length;
        if (dn === 0) return;
        const panelW = Math.min(280, (S.W - 16 * (dn + 1)) / dn);
        const totalW = panelW * dn + 12 * (dn - 1);
        const startX = (S.W - totalW) / 2;
        const panelH = 102;
        const panelY = S.H - panelH - 8;

        for (let i = 0; i < dn; i++) {
            const x = startX + i * (panelW + 12);
            this._drawPanel(ctx, displayPlayers[i], x, panelY, panelW, panelH, mode);
        }
    }

    _renderTimer(ctx, t) {
        const mins = Math.floor(t / 60);
        const secs = Math.floor(t % 60);
        ctx.save();
        ctx.font      = '16px Arial';
        ctx.fillStyle = 'rgba(255,255,255,0.45)';
        ctx.textAlign  = 'center';
        ctx.textBaseline = 'top';
        ctx.fillText(`${mins}:${secs.toString().padStart(2, '0')}`, S.W / 2, 6);
        ctx.restore();
    }

    _renderWaveHeader(ctx, waveNum, enemiesLeft) {
        ctx.save();
        ctx.textAlign    = 'center';
        ctx.textBaseline = 'top';

        // Wave number
        ctx.font      = 'bold 22px Arial';
        ctx.fillStyle = '#ff6644';
        ctx.shadowColor = '#000';
        ctx.shadowBlur  = 6;
        ctx.fillText(`WAVE ${waveNum}`, S.W / 2, 28);
        ctx.shadowBlur = 0;

        // Enemies remaining
        ctx.font      = '14px Arial';
        ctx.fillStyle = '#ccc';
        ctx.fillText(`Enemies remaining: ${enemiesLeft}`, S.W / 2, 54);

        ctx.restore();
    }

    _renderDraftHeader(ctx, remaining) {
        ctx.save();
        ctx.font      = 'bold 18px Arial';
        ctx.textAlign = 'center';
        ctx.textBaseline = 'middle';
        ctx.fillStyle = '#ffd700';
        ctx.shadowColor = '#000';
        ctx.shadowBlur  = 6;
        ctx.fillText('DRAFT MODE', S.W / 2, 22);
        ctx.shadowBlur = 0;
        ctx.font      = '14px Arial';
        ctx.fillStyle = '#4488ff';
        ctx.fillText(`P1: ${remaining[0]} left`, S.W / 2 - 80, 44);
        ctx.fillStyle = '#ff4444';
        ctx.fillText(`P2: ${remaining[1]} left`, S.W / 2 + 80, 44);
        ctx.restore();
    }

    _drawPanel(ctx, player, x, y, w, h, mode) {
        const f     = player.fighter;
        const port  = player.port;
        const color = S.P_COLORS[port % 4];

        ctx.save();

        // ── Panel background ─────────────────────────────────────
        ctx.fillStyle = f.isAlive ? 'rgba(0,0,0,0.7)' : 'rgba(0,0,0,0.4)';
        ctx.beginPath();
        ctx.roundRect(x, y, w, h, 8);
        ctx.fill();

        // Color accent bar (top edge)
        const teamColors = ['#ff4444', '#4488ff', '#44dd44', '#ddaa22'];
        const accentColor = (mode === 'team' && f.team >= 0)
            ? teamColors[f.team] || color
            : (f.isAlive ? color : '#555');
        ctx.fillStyle = accentColor;
        ctx.fillRect(x + 4, y, w - 8, 3);

        // Team label (if team mode)
        if (mode === 'team' && f.team >= 0) {
            const teamLabels = ['TEAM A', 'TEAM B', 'TEAM C', 'TEAM D'];
            ctx.font      = 'bold 10px Arial';
            ctx.fillStyle = teamColors[f.team] || '#fff';
            ctx.textAlign = 'right';
            ctx.fillText(teamLabels[f.team] || 'TEAM ?', x + w - 8, y + 14);
        }

        // ── Player label + character name ─────────────────────────
        ctx.font      = 'bold 14px Arial';
        ctx.fillStyle = color;
        ctx.textAlign = 'left';
        ctx.fillText(`P${port + 1}`, x + 8, y + 20);

        ctx.font      = '11px Arial';
        ctx.fillStyle = player.isAI ? '#777' : '#999';
        ctx.fillText(player.isAI ? 'CPU' : '', x + 32, y + 20);

        ctx.font      = '12px Arial';
        ctx.fillStyle = '#bbb';
        ctx.textAlign = 'right';
        ctx.fillText(f.data.name || '', x + w - 8, y + 20);

        // ── Eliminated state ──────────────────────────────────────
        if (!f.isAlive) {
            ctx.font      = 'bold 28px Arial';
            ctx.fillStyle = '#666';
            ctx.textAlign = 'center';
            ctx.fillText('K.O.', x + w / 2, y + h / 2 + 10);
            ctx.restore();
            return;
        }

        // ── Damage % / Stamina HP ─────────────────────────────────
        if (mode === 'stamina' && f.maxStaminaHP > 0) {
            // HP bar
            const hpW  = w - 24;
            const hpH  = 10;
            const hpX  = x + 12;
            const hpY  = y + 42;
            const ratio = Math.max(0, f.staminaHP / f.maxStaminaHP);

            // Track
            ctx.fillStyle = '#333';
            ctx.beginPath();
            ctx.roundRect(hpX, hpY, hpW, hpH, 4);
            ctx.fill();

            // Fill
            let hpColor;
            if      (ratio > 0.5) hpColor = '#44dd44';
            else if (ratio > 0.25) hpColor = '#ddaa22';
            else                   hpColor = '#dd3333';
            if (ratio > 0) {
                ctx.fillStyle = hpColor;
                ctx.beginPath();
                ctx.roundRect(hpX, hpY, hpW * ratio, hpH, 4);
                ctx.fill();
            }

            // Border
            ctx.strokeStyle = '#666';
            ctx.lineWidth   = 1;
            ctx.beginPath();
            ctx.roundRect(hpX, hpY, hpW, hpH, 4);
            ctx.stroke();

            // HP text
            ctx.font      = 'bold 18px Arial';
            ctx.fillStyle = '#fff';
            ctx.textAlign = 'center';
            ctx.fillText(`${Math.ceil(f.staminaHP)} HP`, x + w / 2, y + 68);
        } else {
            // Standard damage %
            const pct = f.damagePercent;
            let pctColor;
            if      (pct < 50)  pctColor = '#fff';
            else if (pct < 100) pctColor = '#ffdd44';
            else if (pct < 150) pctColor = '#ff8822';
            else                pctColor = '#ff3333';

            ctx.font      = 'bold 36px Arial';
            ctx.fillStyle = pctColor;
            ctx.textAlign = 'center';
            ctx.fillText(`${Math.floor(pct)}%`, x + w / 2, y + 58);
        }

        // ── Stocks (dots) ─────────────────────────────────────────
        const stockY  = y + 78;
        const maxDots = Math.min(f.stocks, 10);
        const dotX    = x + 12;
        for (let s = 0; s < maxDots; s++) {
            ctx.fillStyle = color;
            ctx.beginPath();
            ctx.arc(dotX + s * 14, stockY, 4.5, 0, Math.PI * 2);
            ctx.fill();
        }
        if (f.stocks > 10) {
            ctx.font      = '10px Arial';
            ctx.fillStyle = '#999';
            ctx.textAlign = 'left';
            ctx.fillText(`+${f.stocks - 10}`, dotX + 10 * 14, stockY + 3);
        }

        // ── Ultimate meter ────────────────────────────────────────
        const mw   = 100;
        const mh   = 8;
        const mx   = x + w - mw - 10;
        const my   = stockY - mh / 2;
        const full = f.ultimateMeter >= S.ULT_MAX;

        // Track background
        ctx.fillStyle = '#2a2a2a';
        ctx.beginPath();
        ctx.roundRect(mx, my, mw, mh, 3);
        ctx.fill();

        // Fill
        const fillW = mw * Math.min(1, f.ultimateMeter / S.ULT_MAX);
        if (full) {
            const pulse = 0.7 + 0.3 * Math.sin(this._frame * 0.15);
            ctx.fillStyle   = `rgba(255,215,0,${pulse})`;
            ctx.shadowColor = '#ffd700';
            ctx.shadowBlur  = 8;
        } else {
            ctx.fillStyle = '#5588bb';
        }
        if (fillW > 0) {
            ctx.beginPath();
            ctx.roundRect(mx, my, fillW, mh, 3);
            ctx.fill();
        }
        ctx.shadowBlur = 0;

        // Border
        ctx.strokeStyle = full ? '#ffd700' : '#555';
        ctx.lineWidth   = 1;
        ctx.beginPath();
        ctx.roundRect(mx, my, mw, mh, 3);
        ctx.stroke();

        // ULT READY label
        if (full) {
            ctx.font      = 'bold 10px Arial';
            ctx.fillStyle = '#ffd700';
            ctx.textAlign = 'left';
            ctx.fillText('ULT READY', mx, my - 3);
        }

        // ── Shield bar (slim, below stocks) ───────────────────────
        const shW = 50, shH = 3;
        const shX = x + 10, shY = stockY + 12;
        ctx.fillStyle = '#222';
        ctx.fillRect(shX, shY, shW, shH);

        const shFill = shW * (f.shieldHP / S.SHIELD_MAX_HP);
        ctx.fillStyle = f.shieldHP < 30 ? '#ff4444' : '#44aaff';
        ctx.fillRect(shX, shY, shFill, shH);

        ctx.restore();
    }
}

SMASH.HUD = HUD;
})();

File: engine/hud.py
"""Heads-up display — damage %, stocks, ultimate meter."""

from __future__ import annotations
from typing import List, TYPE_CHECKING
import math

import pygame

from settings import (
    SCREEN_WIDTH, SCREEN_HEIGHT, COLOR_WHITE, PLAYER_COLORS,
    ULTIMATE_METER_MAX,
)

if TYPE_CHECKING:
    from engine.game import Player


class HUD:
    """Renders player info panels at the bottom of the screen."""

    PANEL_HEIGHT = 100
    PANEL_MARGIN = 10
    METER_HEIGHT = 8
    METER_WIDTH = 120

    def __init__(self):
        self.font_large: pygame.font.Font | None = None
        self.font_small: pygame.font.Font | None = None

    def init_fonts(self) -> None:
        self.font_large = pygame.font.SysFont("Arial", 36, bold=True)
        self.font_small = pygame.font.SysFont("Arial", 16)

    def render(self, screen: pygame.Surface, players: List[Player]) -> None:
        if self.font_large is None:
            self.init_fonts()

        num = len(players)
        panel_w = min(250, (SCREEN_WIDTH - self.PANEL_MARGIN * (num + 1)) // num)
        total_w = panel_w * num + self.PANEL_MARGIN * (num - 1)
        start_x = (SCREEN_WIDTH - total_w) // 2
        panel_y = SCREEN_HEIGHT - self.PANEL_HEIGHT - self.PANEL_MARGIN

        for i, player in enumerate(players):
            x = start_x + i * (panel_w + self.PANEL_MARGIN)
            self._render_panel(screen, player, x, panel_y, panel_w)

    def _render_panel(self, screen: pygame.Surface, player: Player,
                      x: int, y: int, w: int) -> None:
        color = PLAYER_COLORS[player.port % len(PLAYER_COLORS)]

        # Background
        panel_surf = pygame.Surface((w, self.PANEL_HEIGHT), pygame.SRCALPHA)
        panel_surf.fill((0, 0, 0, 160))
        pygame.draw.rect(panel_surf, color, (0, 0, w, self.PANEL_HEIGHT), 2)
        screen.blit(panel_surf, (x, y))

        # Player label
        label = self.font_small.render(
            f"P{player.port + 1} {'AI' if player.is_ai else 'Human'}",
            True, color,
        )
        screen.blit(label, (x + 8, y + 4))

        # Damage %
        fighter = player.fighter
        pct_str = f"{fighter.damage_percent:.0f}%"
        # Color ramp: white → yellow → orange → red
        pct = fighter.damage_percent
        if pct < 50:
            pct_color = COLOR_WHITE
        elif pct < 100:
            t = (pct - 50) / 50
            pct_color = (255, int(255 - 35 * t), int(255 - 215 * t))
        elif pct < 150:
            t = (pct - 100) / 50
            pct_color = (255, int(220 - 70 * t), int(40 - 10 * t))
        else:
            pct_color = (220, 50, 50)

        pct_surf = self.font_large.render(pct_str, True, pct_color)
        screen.blit(pct_surf, (x + w // 2 - pct_surf.get_width() // 2, y + 22))

        # Stocks
        stock_y = y + 65
        stock_radius = 6
        for s in range(fighter.stocks):
            pygame.draw.circle(
                screen, color,
                (x + 20 + s * (stock_radius * 2 + 4), stock_y),
                stock_radius,
            )

        # Ultimate meter bar
        meter_x = x + w - self.METER_WIDTH - 10
        meter_y = y + 65 - self.METER_HEIGHT // 2
        # Background
        pygame.draw.rect(screen, (60, 60, 60),
                         (meter_x, meter_y, self.METER_WIDTH, self.METER_HEIGHT))
        # Fill
        fill_w = int(self.METER_WIDTH * (fighter.ultimate_meter / ULTIMATE_METER_MAX))
        bar_color = (255, 215, 0) if fighter.ultimate_meter >= ULTIMATE_METER_MAX else (100, 180, 255)
        pygame.draw.rect(screen, bar_color,
                         (meter_x, meter_y, fill_w, self.METER_HEIGHT))
        # Border
        pygame.draw.rect(screen, COLOR_WHITE,
                         (meter_x, meter_y, self.METER_WIDTH, self.METER_HEIGHT), 1)

        # "ULT" label when ready or cooldown countdown
        if fighter.ultimate_cooldown > 0:
            secs = int(math.ceil(fighter.ultimate_cooldown))
            ult_label = self.font_small.render(f"ULT CD: {secs}s", True, (255, 200, 100))
            screen.blit(ult_label, (meter_x, meter_y - 16))
        elif fighter.ultimate_meter >= ULTIMATE_METER_MAX:
            ult_label = self.font_small.render("ULT READY", True, (255, 215, 0))
            screen.blit(ult_label, (meter_x, meter_y - 16))
